crates/bevy_ecs/src/query/iter.rs

```diff
@@ -153,3 +254,143 @@ impl<'w, 's, Q: WorldQuery> ExactSizeIterator for QueryIter<'w, 's, Q, ()> {
             .sum()
     }
 }
+
+struct QueryIterationCursor<'s, Q: WorldQuery, F: WorldQuery> {
+    table_id_iter: std::slice::Iter<'s, TableId>,
+    archetype_id_iter: std::slice::Iter<'s, ArchetypeId>,
+    fetch: Q::Fetch,
+    filter: F::Fetch,
+    current_len: usize,
+    current_index: usize,
+    is_dense: bool,
+}
+
+impl<'s, Q: WorldQuery, F: WorldQuery> Clone for QueryIterationCursor<'s, Q, F>
+where
+    Q::Fetch: Clone,
+    F::Fetch: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            table_id_iter: self.table_id_iter.clone(),
+            archetype_id_iter: self.archetype_id_iter.clone(),
+            fetch: self.fetch.clone(),
+            filter: self.filter.clone(),
+            current_len: self.current_len,
+            current_index: self.current_index,
+            is_dense: self.is_dense,
+        }
+    }
+}
+
+impl<'s, Q: WorldQuery, F: WorldQuery> QueryIterationCursor<'s, Q, F>
+where
+    F::Fetch: FilterFetch,
+{
+    unsafe fn init_empty(
+        world: &World,
+        query_state: &'s QueryState<Q, F>,
+        last_change_tick: u32,
+        change_tick: u32,
+    ) -> Self {
+        QueryIterationCursor {
+            table_id_iter: [].iter(),
+            archetype_id_iter: [].iter(),
+            ..Self::init(world, query_state, last_change_tick, change_tick)
+        }
+    }
+
+    unsafe fn init(
+        world: &World,
+        query_state: &'s QueryState<Q, F>,
+        last_change_tick: u32,
+        change_tick: u32,
+    ) -> Self {
+        let fetch = <Q::Fetch as Fetch>::init(
+            world,
+            &query_state.fetch_state,
+            last_change_tick,
+            change_tick,
+        );
+        let filter = <F::Fetch as Fetch>::init(
+            world,
+            &query_state.filter_state,
+            last_change_tick,
+            change_tick,
+        );
+        QueryIterationCursor {
+            is_dense: fetch.is_dense() && filter.is_dense(),
+            fetch,
+            filter,
+            table_id_iter: query_state.matched_table_ids.iter(),
+            archetype_id_iter: query_state.matched_archetype_ids.iter(),
+            current_len: 0,
+            current_index: 0,
+        }
+    }
+
+    /// retreive last returned item again
+    #[inline]
+    unsafe fn peek_last<'w>(&mut self) -> Option<<Q::Fetch as Fetch<'w>>::Item> {
+        if self.current_index > 0 {
+            Some(self.fetch.table_fetch(self.current_index - 1))
+        } else {
+            None
+        }
+    }
```

Comment:
Could you add a test for using iter_combinations on queries over components stored in sparse sets? Using `table_fetch` here seems a bit suspicious and I'm not 100% sure this will work how you want it to when the StorageType is set to sparse sets

Reply:
Yeah, setting A's storage to sparse set result in access violation. I'm not yet sure why exactly, as I haven't dug into that part of the code enough yet. I thought that `table_fetch` calls are not not actually mutating any state, thus being idempotent. Clearly I'm wrong.

Reply:
Oh I've figured it out. It is just a wrong method to call for sparse sets. I've missed that.

Reply:
Fixed and test case added.